### ISSUE TITLE
 [GlobalStep] Validation fails to display for Coupons which are restricted for "Allowed emails" using an asterisk "*" to match all parts of emails.

### ISSUE THREAD

**Issue by gglobalstep** [woocommerce/woocommerce#30434]
## Bug Description:
Validation is not displayed for Coupons which are restricted using Allowed emails when they set using an asterisk (*) to match parts of an email like "*@gmail.com" under Users restrictions.

## Environment:
Woocommerce Version : WooCommerce 5.6.0 RC 1 and 5.5.2
WordPress version: v5.8

PC:
Windows 10, Mac 10.14.6
Chrome(Version 91.0.4472.77) 
Firefox(Version 89.0.2)
Safari: v14.0

## Steps To Reproduce:
1. Create any test site using JN site.
1. Install and activate all the required plugins.
1. Complete the OBW setup.
1. Add some products.
1. Create a Coupon with any type.
1. Edit the Coupon and select the Users restriction tab.
1. Click on Allowed emails option and use an asterisk (*) to match parts of an email. for ex. *@gmail.com.
1. Update the Coupon.
1. Add any product to the Cart and go to Checkout page.
1. Enter a different email on billing address like Yahoo or hotmail.
1. Apply the Coupon.
1. Click on Place order button.
1. Observe that no validation is displayed and order gets placed.

## Actual Result:
Product titles are displayed with HTML tags on the Recent comments section on the frontend.

## Expected Result:
Product titles should be displayed as "Bold" rather than with the HTML tags.

## Video:

https://user-images.githubusercontent.com/41110392/128376757-3d1f8caa-7b6b-4079-8b9d-6f7d1cf810a0.mp4

*Isolating the problem (mark completed items with an [x]):*
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

<details>

`
### WordPress Environment ###

WC Version: 5.6.0
REST API Version: ✔ 5.6.0
WC Blocks Version: ✔ 5.5.1
Action Scheduler Version: ✔ 3.2.1
WC Admin Version: ✔ 2.5.0-rc.1
Log Directory Writable: ✔
WP Version: 5.8
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.48 (Unix) OpenSSL/1.0.2g
PHP Version: 7.4.21
PHP Post Max Size: 1 GB
PHP Time Limit: 30
PHP Max Input Vars: 5000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.33-0ubuntu0.16.04.1-log
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 5.6.0
WC Database Prefix: wp_
Total Database Size: 11.73MB
Database Data Size: 8.97MB
Database Index Size: 2.76MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.08MB + Index: 0.13MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_gla_budget_recommendations: Data: 0.22MB + Index: 0.14MB + Engine InnoDB
wp_gla_merchant_issues: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_gla_shipping_rates: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_gla_shipping_times: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailchimp_carts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailchimp_jobs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_custom_fields: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_dynamic_segment_filters: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_feature_flags: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_forms: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_log: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_mapping_to_external_entities: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletters: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_newsletter_links: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_mailpoet_newsletter_option: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_option_fields: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_posts: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_segment: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_templates: Data: 2.52MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_scheduled_tasks: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_scheduled_task_subscribers: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_segments: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_sending_queues: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_settings: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_statistics_clicks: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_mailpoet_statistics_forms: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_statistics_newsletters: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_statistics_opens: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_mailpoet_statistics_unsubscribes: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_mailpoet_statistics_woocommerce_purchases: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_mailpoet_stats_notifications: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_subscribers: Data: 0.02MB + Index: 0.13MB + Engine InnoDB
wp_mailpoet_subscriber_custom_field: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_subscriber_ips: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_subscriber_segment: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_user_flags: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_ms_snippets: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_options: Data: 4.03MB + Index: 0.16MB + Engine InnoDB
wp_postmeta: Data: 0.17MB + Index: 0.14MB + Engine InnoDB
wp_posts: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
wp_snippets: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wpml_mails: Data: 0.27MB + Index: 0.00MB + Engine InnoDB

### Post Type Counts ###

attachment: 24
customize_changeset: 6
mailpoet_page: 1
nav_menu_item: 5
page: 11
post: 2
product: 28
product_variation: 25
revision: 6
shop_coupon: 1
shop_order: 8

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (11) ###

Query Monitor: by John Blackbourn – 3.7.1
Code Snippets: by Code Snippets Pro – 2.14.1
Companion Plugin: by Osk – 1.18
Google Listings and Ads: by WooCommerce – 1.3.0
Jetpack: by Automattic – 10.0
Mailchimp for WooCommerce: by Mailchimp – 2.5.2
MailPoet 3 (New): by MailPoet – 3.65.1
WooCommerce Beta Tester: by WooCommerce – 2.0.2
WooCommerce PayPal Payments: by WooCommerce – 1.4.0
WooCommerce: by Automattic – 5.6.0-rc.1
WP Mail Logging: by Wysija – 1.9.8

### Inactive Plugins (4) ###

Akismet Anti-Spam: by Automattic – 4.1.10
Gutenberg: by Gutenberg Team – 11.2.0
Hello Dolly: by Matt Mullenweg – 1.7.2
WooCommerce Shipping & Tax: by WooCommerce – 1.25.17

### Dropin Plugins (1) ###

db.php: Query Monitor Database Class

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 3.7.0
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 186
Oldest: 2021-08-04 08:07:34 +0000
Newest: 2021-08-05 14:27:32 +0000

### Status report information ###

Generated at: 2021-08-05 15:06:19 +00:00
`

</details>

**Comment by juliaamosova** [comment]
Hi @gglobalstep,

Thank you for taking the time to report this bug, we really appreciate your help. I can reproduce it on my end using WooCommerce 5.5.2 + WP 5.8. 

My coupon set-up:

![coupon](https://user-images.githubusercontent.com/19143190/128564795-040afe13-7abf-43ac-997d-ed8b54cfccf8.png)

Placing an order applying the coupon and using non-gmail email - order went through:

http://g.recordit.co/fMHRAwkhXJ.gif

Priority of this issue will be assigned shortly after we'd go through a round of prioritization. Please note it may take a couple of days.

**Comment by roykho** [comment]
Hi there!

So I looked into this and the reason you are able to proceed with the checkout is because the customer/user account you're using has a `gmail.com` email associated with it. If you check your user contact info email, you probably have a `gmail.com` email for this user. If you change this and re-test you can see it will no longer let you checkout.

Then looking at the code, it would appear this is by design. You can see it explicitly checks for both emails, the user's contact info email and the billing address email. https://github.com/woocommerce/woocommerce/blob/trunk/includes/class-wc-cart.php#L1631-L1632

@juliaamosova - what do you think?

**Comment by juliaamosova** [comment]
@roykho thanks for digging in here! 

I confirm that I had a user email set up as a `gmail` one indeed when I tested originally. 

I just tested again by changing the user email to a non-gmail one and was no longer able to checkout using the `yahoo` email address as specified in [my comment above](https://github.com/woocommerce/woocommerce/issues/30434#issuecomment-894487089). I am getting an error message instead as expected:

<img width="836" alt="coupon" src="https://user-images.githubusercontent.com/19143190/128924838-022f78b3-44ad-4aee-96d7-0a81d763105a.png">
 
With this in mind, I think we can close this issue as things are working as expected according to the code reference you provided. I'll go ahead and close this. Thanks again!
